### PR TITLE
GH 15 Explore initial design of the `cardano-shell` architecture

### PR DESCRIPTION
**Pull request by ksaric**
https://github.com/input-output-hk/cardano-shell/issues/15

I consider this to be the "very best of", containing hits such as "Async doesn't compose", "Concurrency is Tricky" and others.

The result from running the shell:
```
Starting up logging feature!
Starting up networking feature!
Running node/wallet/whatever!
Running node/wallet/whatever!
Running node/wallet/whatever!
Running node/wallet/whatever!
Running node/wallet/whatever!
Exception occured!
Shutting down networking feature!
Shutting down logging feature!
```

### CODE REVIEW

File: src/Lib.hs [hunk 1]
```diff
@@ -1,30 +1,110 @@
...
+
+    -- A general pattern. The dependency is always in a new thread, and we depend on it,
+    -- so when that dependency gets shut down all the other features that depend on it get
+    -- shut down as well.
+    -- http://hackage.haskell.org/package/async-2.2.1/docs/Control-Concurrent-Async.html#v:withAsync
+    -- http://hackage.haskell.org/package/base-4.12.0.0/docs/Control-Concurrent.html#v:forkFinally
+    let noDependency :: IO NoDependency
+        noDependency = pure NoDependency
+
+    _ <- withFeatureUsing noDependency $ \loggingDependency -> do
```
Comment (dcoutts) [inline src/Lib.hs]: You don't need this one surely.

File: src/Types.hs [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+-- | The construction same as/very similar to @Async@ which allows us to cancel specific threads/features.
+withFeatureUsing :: forall a b. IO a -> (a -> IO b) -> IO b
+withFeatureUsing = \action inner -> do
+  var <- newEmptyMVar
+
+  mask $ \restore -> do
+    -- In another thread, we don't know the type of exception beforehand.
+    threadId    <- forkIO $ try @SomeException (restore action) >>= putMVar var
+    var         <- readMVar var
```
Comment (dcoutts) [inline src/Types.hs]: This is running the action synchronously. There's no async running of the feature at all.

File: src/Features/Networking.hs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+    featureCleanup' :: NetworkLayer -> IO ()
+    featureCleanup' _ = putTextLn "Shutting down networking feature!" -- close all connections, for example
+
+-- MonadConc m => m a -> m (Async m a)
+-- :: forall m. (MonadConc m, MonadIO m) => CardanoEnvironment -> Async LoggingLayer -> CardanoConfiguration -> Text -> m (Async m NetworkLayer)
+actualNetworkFeature :: CardanoEnvironment -> LoggingLayer -> CardanoConfiguration -> Text -> IO NetworkLayer
+actualNetworkFeature _ asyncLoggingLayer _ _ = do
+    putTextLn "Starting up networking feature!"
+    pure $ testNetworkLayer asyncLoggingLayer
```
Comment (dcoutts) [inline src/Features/Networking.hs]: Note how this is a synhronous interface now. The feature cannot continue to run here unless it does its own forkIO.

**Comment by ksaric** [comment]
Closing in favor of https://github.com/input-output-hk/cardano-shell/pull/17

**Pull request closed without merging**